### ISSUE TITLE
Tunnel UDP

### ISSUE THREAD

**Issue by cornzzy** [neevek/omnip#5]
Is UDP over QUIC available? Any way to tunnel wireguard on this setup?

`client` -> `server1 omni client` -> `server2 omni and wg server`

**Comment by neevek** [comment]
Currently UDP tunneling is not supported, I am working on something related to that though.
I am curious about your use case, can you elaborate on that a little bit?

**Comment by cornzzy** [comment]
Wireguard directly is blocked and it uses UDP. I use `server1` as a middle local server which listens on UDP from client wireguard and sends it as QUIC to abroad `server2`. [udp2raw](https://github.com/wangyu-/udp2raw) does it but it's not QUIC.

**Comment by neevek** [comment]
> Wireguard directly is blocked and it uses UDP. I use `server1` as a middle local server which listens on UDP from client wireguard and sends it as QUIC to abroad `server2`. [udp2raw](https://github.com/wangyu-/udp2raw) does it but it's not QUIC.

Well this will be supported soon, as the UDP over QUIC part is already implemented in the rstun project, which omnip depends on to support all the over-QUIC tunneling.

**Comment by cornzzy** [comment]
I checked rstun first. Are you saying I can I use rstun right now for UDP? I thought it only handles TCP. This part of the readme is why I made this issue:
![image](https://github.com/user-attachments/assets/857c89f8-12f6-46c0-bb49-956b5c7a7c28)

**Comment by neevek** [comment]
> Does that mean I can use rstun right now for UDP? I thought it only handles TCP. This part of the readme is why I made this issue: ![image](https://private-user-images.githubusercontent.com/116248852/371824403-857c89f8-12f6-46c0-bb49-956b5c7a7c28.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc2OTc1NjUsIm5iZiI6MTcyNzY5NzI2NSwicGF0aCI6Ii8xMTYyNDg4NTIvMzcxODI0NDAzLTg1N2M4OWY4LTEyZjYtNDZjMC1iYjQ5LTk1NmI1YzdhN2MyOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkzMFQxMTU0MjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYTVjOGEwMzBlYWJjOTRhYWQ3YjczOWJhYTAzM2NlODE5YTI2ZTFkMmJkN2FlNWM2YWQzM2M1MjcxODkxZTM1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.faQptfdr1EgTfeRZIWk8G3Hc8Vf48_sknVTG_nWA-GE)

The code that implements UDP tunneling is not merged into master branch yet, the README is also not updated.
Anyway I just released a beta version that includes the feature, check it out [here](https://github.com/neevek/rstun/releases/tag/release%2F0.6.0-beta) and give it a try, it will be great if you can report back later.

**Comment by cornzzy** [comment]
It seems like QUIC is blocked on my "internet". These can affect it:
- set fingerprint, for example chrome, firefox or google apps.
- set SNI

If this is in your interest, I can try again with different fingerprint and SNI.
It could also be wireguard MTU. What MTU should I set? Currently it's 1420.
Here is the log, tunnel is created but packets drop on strict internet.

![image](https://github.com/user-attachments/assets/f8e4e856-2999-42db-971f-963b0f5140e2)

**Comment by neevek** [comment]
> It seems like QUIC is blocked on my "internet". These can affect it:
> - set fingerprint, for example chrome, firefox or google apps.
> - set SNI
> 
> If this is in your interest, I can try again with different fingerprint and SNI.
> It could also be wireguard MTU. What MTU should I set? Currently it's 1420.
> Here is the log, tunnel is created but packets drop on strict internet.
> 
> ![image](https://github.com/user-attachments/assets/f8e4e856-2999-42db-971f-963b0f5140e2)
> 

I think it is most probably related to MTU, try setting it to 1200.

**Comment by cornzzy** [comment]
The lowest MTU wg accepts is 1280, same log happened with 1280. Also tried shadowsocks TCP with the previous version and the same things happens.

**Comment by neevek** [comment]
@cornzzy check out this [one](https://github.com/neevek/rstun/releases/tag/release%2F0.6.0), and if possible let me know if it works as expected.

**Comment by cornzzy** [comment]
Awesome, it works on 1420 MTU and with great speed. I'm going to put 40 clients on it and see if it gets detected and blocked by DPI.

There are some warnings but I don't know if its a problem.
`2024-10-14 00:39:28.228 [rstunc] [rstun::udp::udp_tunnel:145] [W] failed to read for udp, err: unexpected end of file`
`2024-10-14 04:21:13.228 [rstund] [rstun::udp::udp_tunnel:274] [W] failed to read from udp packet from tunnel, err: unexpected end of file`

As a side note, it would be great if dns was moved to options like `--dns 9.9.9.9`. This is too opinionated:
![image](https://github.com/user-attachments/assets/fc626b54-0263-46c4-8113-868f22957558)

**Comment by neevek** [comment]
> Awesome, it works on 1420 MTU and with great speed. I'm going to put 40 clients on it and see if it gets detected and blocked by DPI.

Good to hear that. By the way, what kind of DPI system you are testing against?

> As a side note, it would be great if dns was moved to options like `--dns 9.9.9.9` or `--dns https://dns.quad9.net/dns-query`. This is too opinionated: ![image](https://private-user-images.githubusercontent.com/116248852/376111088-fc626b54-0263-46c4-8113-868f22957558.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg5MDY5NjksIm5iZiI6MTcyODkwNjY2OSwicGF0aCI6Ii8xMTYyNDg4NTIvMzc2MTExMDg4LWZjNjI2YjU0LTAyNjMtNDZjNC04MTEzLTg2OGYyMjk1NzU1OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDE0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAxNFQxMTUxMDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NGExOWEyN2ZhYWI0ZGVhMTc1MGM2YmI3NTUyZmM4NjA4YmY2YWQ3ZDFlYzkyODg3MTZmZDNhN2JmNDAzN2E2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jDRQTkt0Tb-wr9N6HY0fYsz1Mus4Tn_sGoJyPIcTFtI)

Sure, I will fix that in later release.

**Comment by cornzzy** [comment]
Government censorships (Ir‌‌a‌‌‌‌n), but I believe it's the same as C‌hi‌n‌‌‌‌a if not worse.

I tried many tunnels such as chisel (ssh), wstunnel (websocket), pingtunnel (icmp) and they all get blocked within days. Some xray/v2ray stuff and udp2raw work but bad performance. Let's see how this one goes 🫡

**Comment by cornzzy** [comment]
Yes it went out. It was worth trying.
It had the highest speed I've seen while it lasted.

![image](https://github.com/user-attachments/assets/3a1b1ad2-3400-4130-9cc1-45a0301af31a)
![image](https://github.com/user-attachments/assets/b6f0aef5-8461-4f19-ab27-558cd12f19d3)
![image](https://github.com/user-attachments/assets/55c58cc1-af39-4b0a-9f7a-a881feb6da8b)
![image](https://github.com/user-attachments/assets/de8eb52a-bc36-4d88-88e4-dd4d459e9c75)

**Comment by cornzzy** [comment]
Yes it went out. It was worth trying.
It had the highest speed I've seen while it lasted.

![image](https://github.com/user-attachments/assets/7ca5af2c-93d2-4bf9-a37a-5f2af04b8961)
![image](https://github.com/user-attachments/assets/1f76be4f-3746-4216-962c-e097b7249b89)
![image](https://github.com/user-attachments/assets/396feeab-290c-45fc-a795-2cee0f0b9068)

**Comment by neevek** [comment]
@cornzzy udp tunneling is supported in [latest release](https://github.com/neevek/omnip/releases/tag/release%2F0.6.0) of omnip with `udp://` and `udp+quic://`.
